File: runtime/ruby/units/wasi/clock_time_get.rb
# requires: memory/ids, rt/m64
def wasi_clock_time_get(id, _precision, out_ptr)
  ns =
    case id
    when 0 then Process.clock_gettime(Process::CLOCK_REALTIME, :nanosecond)
    when 1, 2, 3 then Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
    else return ERRNO_INVAL
    end
  @memory.ids(out_ptr, Rt.m64(ns))
  ERRNO_SUCCESS
end

File: runtime/python/units/wasi/clock_time_get.py
# requires: memory/i64_store
def wasi_clock_time_get(self, id, precision, out_ptr):
    if id == 0:  # realtime
        ns = time.time_ns()
    elif id in (1, 2, 3):  # monotonic / process / thread cputime
        ns = time.monotonic_ns()
    else:
        return self.ERRNO_INVAL
    self.memory.i64_store(out_ptr, ns & Rt.M64)
    return self.ERRNO_SUCCESS
